app/bitbox02-api.d.ts

```diff
@@ -0,0 +1,105 @@
+declare module 'bitbox02-api' {
+  export function getDevicePath(options?: {forceBridge: boolean}): string
+
+  type CardanoNetwork = 0 | 1
+
+  export const constants = {
+    Status: {
+      PairingFailed: string,
+    },
+    Product: {
+      BitBox02Multi: string,
+      BitBox02BTCOnly: string,
+    },
+    messages: {
+      CardanoNetwork: {
+        CardanoMainnet: CardanoNetwork,
+        CardanoTestnet: CardanoNetwork,
+      },
+    },
+  }
+
+  declare class Firmware {
+    Product(): string
+  }
+
+  type Keypath = number[]
+  type ScriptConfig = {
+    pkhSkh: {
+      keypathPayment: Keypath
+      keypathStake: Keypath
+    }
+  }
+
+  type CardanoInput = {
+    keypath: Keypath
+    prevOutHash: Uint8Array
+    prevOutIndex: number
+  }
+
+  type CardanoOutput = {
+    encodedAddress: string
+    value: string
+    scriptConfig?: ScriptConfig
+  }
+
+  type CardanoCertificate =
+    | {
+        stakeRegistration: {
+          keypath: Keypath
+        }
+      }
+    | {
+        stakeDeregistration: {
+          keypath: Keypath
+        }
+      }
+    | {
+        stakeDelegation: {
+          keypath: Keypath
+          poolKeyhash: Uint8Array
+        }
+      }
+
+  type CardanoWithdrawal = {
+    keypath: Keypath
+    value: string
+  }
+
+  type CardanoShelleyWitness = {
+    signature: Uint8Array
+    publicKey: Uint8Array
+  }
+
+  export declare class BitBox02API {
+    constructor(devicePath: string)
+    connect(
+      showPairingCb: (string) => void,
+      userVerify: () => Promise<void>,
+      handleAttestationCb: (bool) => void,
+      onCloseCb: () => void,
+      setStatusCb: (string) => void
+    )
+    close(): boolean
+    firmware(): Firmware
+    version(): string
+    cardanoXPubs(keypaths: Keypath[]): Promise<Uint8Array[]>
+    cardanoAddress(
+      network: CardanoNetwork,
+      scriptConfig: ScriptConfig,
+      display?: boolean
+    ): Promise<string>
+    cardanoSignTransaction(params: {
+      network: CardanoNetwork
+      inputs: CardanoInput[]
+      outputs: CardanoOutput[]
+      fee: string
+      ttl: string
```

Comment:
@benma btw ttl in cardano transactions is optional since Mary era, similarly as `validityIntervalStart` - we actually didn't implement that in Adalite so far, but I'm making a PR to fix that: https://github.com/vacuumlabs/adalite/pull/1226/files

So alongside multiasset support it would be great if you included also support for optional TTL for the sake of consistency

Reply:
The BitBox02 as well as the bitbox02-api package actually accept undefined/null for ttl and validityIntervalStart and handle it properly. You can simply adjust the type here to `string | null` like with `validityIntervalStart`.

Reply:
I tried passing ttl=null, but the signature returned by bitbox didn't match with the tx body that our code serialized. Do you have some test fixture showing how bitbox serializes a transaction with null ttl? null ttl should be missing completely from the transaction body as per the CDDL

Reply:
We have test fixtures, but none where ttl is zero at the moment.

The relevant cbor serialization code is here:

https://github.com/digitalbitbox/bitbox02-firmware/blob/d7e59db2fde8823354c49b2325b2656992a0e32b/src/rust/bitbox02-rust/src/hww/api/cardano/sign_transaction/cbor.rs#L123-L126

I'll need a bit more time to make a test and compare it to other implementations to see where the root of the problem is.

Adalite always sets a ttl though, right?

Reply:
Adalite UI yes, but the crypto providers are generic and we allow also pool registrations passed as raw CBOR and these can have TTL not defined. BTW, looking at the code you attached - ttl = 0 is in theory a semantically valid TTL the transaction can have - advanced users may leverage that to sign a transaction and ensure it won't ever be accepted by the blockchain - it's used e.g. with Ledger when CLI users need to sign separately voting registration metadata (but the device allows signing only the full tx) and therefore want to throw away the rest of the signed transaction. (of course, they can use ttl=1 which would do as well, but mapping 0 to null seems worth avoiding for the sake of predictability of the interface)

So it would be ideal, if your implementation distinguished between ttl being `null` and `0` and treated `0` as a possible TTL value, to avoid confusion

Reply:
Right, the firmware so far treats ttl=0 the same as if it was missing. Interesting, I wasn't aware that a zero (but not missing) TTL was used with a good reason. I will add this change to the next release, fixing this part.

That does not explain yet why `ttl=null` in your test yielded different serializations by the firmware and the Adalite code. I will investigate and come back to you.

Note that the BitBox02 doesn't support pool registrations (also disabled in the BB02 crypto provider in Adalite), so I think this edge case is not a blocker for release. If the issue turns out to be in the BitBox02 firmware, we will resolve it by the next release.

Reply:
> Note that the BitBox02 doesn't support pool registrations (also disabled in the BB02 crypto provider in Adalite), so I think this edge case is not a blocker for release.

Yes, it isn't a blocker, we are planning to deal with that null ttl issue separately in #1226 :+1:

Reply:
@refi93 

Good news: sending an undefined ttl actually works, and the BitBox02 and Adalite both serialize the tx the same way. I just send an onchain transaction using this branch:

https://github.com/benma/adalite/commit/a58ebbb33a3879b46084b347de602701abdf68f0

Since the type of `ttl` can't currently be null or undefined, I changed the type to be `number | undefined`, which resulted in these two important changes:

a)

Don't serialize the ttl if it is not undefined, not if it is not null. Note that the current check `!== null` does not do anything as in the current type, it cannot be null).

https://github.com/benma/adalite/commit/a58ebbb33a3879b46084b347de602701abdf68f0#diff-32ff170d6209da4ac450ad17009a98373855e630758dba679e4afc7f4b0c6a99R94

b)

Don't do `toString()` without checking for undefined first:

https://github.com/benma/adalite/commit/a58ebbb33a3879b46084b347de602701abdf68f0#diff-3a3f4139e69bc5cbc4f08f613bb13fc943236d3825d2e105dd467f9de58014a8R309

I suspect that one of these changes were missing from your experiment.